python-package/packager/pep517.py

```diff
@@ -0,0 +1,157 @@
+        lib_path = pkg_path / "lib"
+        lib_path.mkdir()
+        libxgboost = locate_or_build_libxgboost(
+            TOPLEVEL_DIR, build_dir=build_dir, build_config=build_config
+        )
+        copy_with_logging(libxgboost, lib_path, logger=logger)
+
+        with cd(workspace):
+            wheel_name = hatchling.build.build_wheel(
+                wheel_directory, config_settings, metadata_directory
+            )
+    return wheel_name
+
+
+def build_sdist(
+    sdist_directory: str,
+    config_settings: Optional[Dict[str, Any]] = None,
+) -> str:
+    """Build a source distribution"""
+    logger = logging.getLogger("xgboost.packager.build_sdist")
+
+    if config_settings:
+        raise NotImplementedError(
+            "XGBoost's custom build backend doesn't support config_settings option "
+            f"when building sdist. {config_settings=}"
+        )
+
+    cpp_src_dir = TOPLEVEL_DIR.parent
+    if not cpp_src_dir.joinpath("CMakeLists.txt").exists():
+        raise RuntimeError(f"Did not find CMakeLists.txt from {cpp_src_dir}")
+
+    # Create tempdir with Python package + C++ sources
+    with tempfile.TemporaryDirectory() as td:
+        td_path = pathlib.Path(td)
+
+        workspace = td_path / "sdist_workspace"
+        workspace.mkdir()
+        logger.info("Copying project files to temporary directory %s", str(workspace))
+
+        copy_with_logging(TOPLEVEL_DIR / "pyproject.toml", workspace, logger=logger)
+        copy_with_logging(TOPLEVEL_DIR / "hatch_build.py", workspace, logger=logger)
+        copy_with_logging(TOPLEVEL_DIR / "README.rst", workspace, logger=logger)
+
+        copytree_with_logging(
+            TOPLEVEL_DIR / "xgboost", workspace / "xgboost", logger=logger
+        )
+        copytree_with_logging(
+            TOPLEVEL_DIR / "packager", workspace / "packager", logger=logger
+        )
+
+        temp_cpp_src_dir = workspace / "cpp_src"
+        copy_cpp_src_tree(cpp_src_dir, target_dir=temp_cpp_src_dir, logger=logger)
+
+        with cd(workspace):
+            sdist_name = hatchling.build.build_sdist(sdist_directory, config_settings)
+    return sdist_name
+
+
+def build_editable(
+    wheel_directory: str,
+    config_settings: Optional[Dict[str, Any]] = None,
+    metadata_directory: Optional[str] = None,
+) -> str:
+    """Build an editable installation. We mostly delegate to Hatchling."""
+    logger = logging.getLogger("xgboost.packager.build_editable")
+
+    if config_settings:
+        raise NotImplementedError(
+            "XGBoost's custom build backend doesn't support config_settings option "
+            f"when building editable installation. {config_settings=}"
+        )
+
+    if locate_local_libxgboost(TOPLEVEL_DIR, logger=logger) is None:
+        raise AssertionError(
+            "To use the editable installation, first build libxgboost with CMake. "
+            "See https://xgboost.readthedocs.io/en/latest/build.html for detailed instructions."
+        )
```

Comment:
Out of curiosity, what happens if you remove this? Is it possible to use an editable installation that also builds libxgboost? In that situation running `pip install -e` would need to rebuild the library as well.

Not saying this feature should be implemented, just want to understand the current setup of the backend to figure out how best to work with this in the future for other RAPIDS libraries too.

Reply:
> what happens if you remove this?

Editable installation won't build `libxgboost.so`, so `import xgboost` will error out.

> Is it possible to use an editable installation that also builds libxgboost?

I decided against it, as there is no reliable way for `packager` module to detect changes in the C++ code. If we want to re-build `libxgboost.so` upon changes made to C++, `scikit-build-core` will probably be a better choice.

I ultimately decided against `scikit-build-core` because:
* It doesn't yet support editable installation.
* I couldn't customize it enough to fit the current use cases of XGBoost.
* Much of what `scikit-learn-core` offers is not useful for XGBoost, since XGBoost does not use compiled extensions. XGBoost uses `ctypes` exclusively to access C API functions from `libxgboost.so`.

The calculus will be different for RAPIDS libraries which uses compiled extensions extensively. IMHO, if you use compiled extensions a lot, it's better to use `scikit-build-core` and give up on editable installation.

Reply:
For what it's worth, `scikit-build-core` just added experimental support for editable installs: https://github.com/scikit-build/scikit-build-core/releases/tag/v0.3.0

Not sure exactly what "experimental" means, to be fair, and to what extent that's acceptable for a project as widely-used as XGBoost. But just wanted to note it.

I am planning to pursue `scikit-build-core` for LightGBM (https://github.com/microsoft/LightGBM/pull/5759), to not have to have any lingering Python glue score that's exposed to the changes in Python packaging. That is still very much a work in progress though, so I'm definitely not recommending that you switch to that over `hatchling`... just noting the link as something you may want to watch as we all figure out the `setup.py`-free world together 😊

Reply:
@jameslamb Guess my knowledge was outdated :) In the future, I'll definitely consider `scikit-learn-core` as a potential alternative.

Reply:
I will be migrating RAPIDS to use `scikit-build-core` at some point in the not-too-distant future. We're already using `scikit-build`, so the transition is the most natural way to get us onto a PEP 517 build. 

Support for editable installs is a must for RAPIDS, and is one of the reasons I haven't attempted the migration yet. [That feature request is something they're aware of](https://github.com/scikit-build/scikit-build-core/issues/114) and are working on. [My understanding is that the current experimental support](https://github.com/scikit-build/scikit-build-core/pull/212) only covers some possible cases around library changes but not all. I will definitely be keeping a close eye on this. I expect full support for editable installs is a precondition of a 1.0 release, and the project has been moving at a reasonable pace so it shouldn't be too long before that exists.

That said, at this stage `scikit-build-core` is definitely less mature than hatchling and I can't fault your choice since you seem to have been able to get everything working fairly painlessly :slightly_smiling_face:

Reply:
@vyasr @hcho3 @trivialfis FYI I have a PR ready over in LightGBM for using `scikit-build-core`: https://github.com/microsoft/LightGBM/pull/5759. If you have time and interested, I'd appreciate any comments you have on it.

We don't care about editable installs for `lightgbm`, and I'm willing to go contribute to `scikit-build-core` (both code and discussions) as we discover issues in it, to help that tool mature.

Reply:
Thanks for the ping! I've started testing the migration for a couple RAPIDS libraries as well, see https://github.com/rapidsai/rmm/pull/1287 and https://github.com/rapidsai/cudf/pull/13531. Editable installs are already in significantly better shape than they were a couple of months ago FWIW.

Reply:
Nice! I learned a lot about `scikit-build-core` in this process, `@` me any time on those RAPIDS PRs if you want an opinion or another set of eyes.